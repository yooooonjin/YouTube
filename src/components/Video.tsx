import { useQuery } from '@tanstack/react-query';
import { useYoutubeApi } from '../context/YoutubeApiContext';
import { DetailVideo } from '../model/video';
import YoutubePlayer from './YoutubePlayer';

type Props = {
  video: DetailVideo;
};

export default function Video({ video }: Props) {
  const { youtube } = useYoutubeApi();

  const { data: channel } = useQuery({
    queryKey: ['channel', video.channelId],
    queryFn: async () => youtube?.getDetailChannel(video.channelId),
    staleTime: 1000 * 60 * 5,
  });

  return (
    <section>
      {!video ? (
        <p>loading...</p>
      ) : (
        <>
          <YoutubePlayer videoId={video.id} />
          <div className='mt-5'>
            <h1 className=' font-semibold mb-5'>{video.title}</h1>
            {channel && (
              <div className='flex items-center mb-5'>
                {channel.thumbnail && (
                  <img
                    className='rounded-full mr-2'
                    src={channel.thumbnail}
                    alt='channel_thumbnail'
                    width={30}
                  />
                )}
                <p>{channel.title}</p>
              </div>
            )}
            <p className='text-xs line-clamp-3'>{video.description}</p>
          </div>
        </>
      )}
    </section>
  );
}
