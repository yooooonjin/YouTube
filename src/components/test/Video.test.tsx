import { render, screen, waitFor } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { withAllContexts, withRouter } from '../../test/utils';
import { fakeDetailVideo as detailVideo } from '../../test/videos';
import Video from '../Video';

describe('Video', () => {
  let fakeYoutube: any = { getDetailChannel: jest.fn() };
  afterEach(() => {
    fakeYoutube.getDetailChannel.mockReset();
  });

  it('renders correctly', async () => {
    const { asFragment } = renderChannelInfoWithCallback(() => ({
      title: 'channel title',
      thumbnail: 'channel thumbnail',
    }));
    await waitFor(() => screen.getByRole('img'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders without URL', async () => {
    renderChannelInfoWithCallback(() => {
      throw new Error('error');
    });
    expect(screen.queryByRole('img')).toBeNull();
  });

  it('renders with URL', async () => {
    renderChannelInfoWithCallback(() => ({
      title: 'channel title',
      thumbnail: 'channel thumbnail',
    }));
    expect(screen.findByRole('img'));
  });

  function renderChannelInfoWithCallback(callback: Function) {
    fakeYoutube.getDetailChannel.mockImplementation(callback);
    return render(
      withAllContexts(
        withRouter(<Route path='/' element={<Video video={detailVideo} />} />),
        fakeYoutube
      )
    );
  }
});
