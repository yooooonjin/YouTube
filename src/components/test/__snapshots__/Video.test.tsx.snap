// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video renders correctly 1`] = `
<DocumentFragment>
  <section>
    <div
      class="w-full"
    >
      <div
        class=""
        id=""
      />
    </div>
    <div
      class="mt-5"
    >
      <h1
        class=" font-semibold mb-5"
      >
        video title
      </h1>
      <div
        class="flex items-center mb-5"
      >
        <img
          alt="channel_thumbnail"
          class="rounded-full mr-2"
          src="channel thumbnail"
          width="30"
        />
        <p>
          channel title
        </p>
      </div>
      <p
        class="text-xs line-clamp-3"
      >
        video description
      </p>
    </div>
  </section>
</DocumentFragment>
`;
